### PR TITLE
Fix: standard material options created shader unexpectly

### LINKED ISSUE #3189: Shader is created unexpectly when clearCoat and clearCoatGlossiness change
### Description

When you change `clearCoat` and `clearCoatGlossiness` in `StandardMaterial`, it will create shader unexpectly.

### Steps to Reproduce

1. Create a blank project.
2. Launch it.
3. Type the following snippet:

```js
console.log('shader count: ', Object.keys(pc.app.graphicsDevice.programLib._cache).length);
for (let i = 0; i < 10; i++) {
  setTimeout(() => {
    let m = pc.app.root.findByName('Box').model.model.meshInstances[0].material;
    m.clearCoat += 0.01;
    m.update();
    console.log('shader count: ', Object.keys(pc.app.graphicsDevice.programLib._cache).length);
  }, i * 100);
}
```

image

You will see it created 10 shaders !

### Reason

In [StandardMaterialOptionsBuilder](https://github.com/playcanvas/engine/blob/master/src/scene/materials/standard-material-options-builder.js#L42-L43):

```diff
StandardMaterialOptionsBuilder.prototype.updateRef = function (options, device, scene, stdMat, objDefs, staticLightList, pass, sortedLights, prefilteredCubeMap128) {
    this._updateSharedOptions(options, stdMat, objDefs, pass);
    options.useTexCubeLod = device.useTexCubeLod;
    this._updateEnvOptions(options, stdMat, scene, prefilteredCubeMap128);
    this._updateMaterialOptions(options, stdMat);
    if (pass === SHADER_FORWARDHDR) {
        if (options.gamma) options.gamma = GAMMA_SRGBHDR;
        options.toneMap = TONEMAP_LINEAR;
    }
    options.hasTangents = objDefs && stdMat.normalMap && ((objDefs & SHADERDEF_TANGENTS) !== 0);
    this._updateLightOptions(options, stdMat, objDefs, sortedLights, staticLightList);
    this._updateUVOptions(options, stdMat, objDefs, false);
+   options.clearCoat = stdMat.clearCoat;
+   options.clearCoatGlossiness = stdMat.clearCoatGlossiness;
};
```

The `options` save `clearCoat` and `clearCoatGlossiness` actual value.

However in [standard.js](https://github.com/playcanvas/engine/blob/master/src/graphics/program-lib/programs/standard.js#L197-L200):

```diff
generateKey: function (options) {
    // ....
    var key = "standard";

    var i;
    for (i = 0; i < props.length; i++) {
        if (options[props[i]])
+            key += props[i] + options[props[i]];
    }

    // ....

    return hashCode(key);
}
```

This means when `clearCoat` and `clearCoatGlossiness` changed, a new shader will be created. This will cause performance issue when doing Interpolation on `clearCoat` and `clearCoatGlossiness`.

### Solution

A quick fix is change `number` to `boolean` in the `options`:

```diff
StandardMaterialOptionsBuilder.prototype.updateRef = function (options, device, scene, stdMat, objDefs, staticLightList, pass, sortedLights, prefilteredCubeMap128) {
    this._updateSharedOptions(options, stdMat, objDefs, pass);
    options.useTexCubeLod = device.useTexCubeLod;
    this._updateEnvOptions(options, stdMat, scene, prefilteredCubeMap128);
    this._updateMaterialOptions(options, stdMat);
    if (pass === SHADER_FORWARDHDR) {
        if (options.gamma) options.gamma = GAMMA_SRGBHDR;
        options.toneMap = TONEMAP_LINEAR;
    }
    options.hasTangents = objDefs && stdMat.normalMap && ((objDefs & SHADERDEF_TANGENTS) !== 0);
    this._updateLightOptions(options, stdMat, objDefs, sortedLights, staticLightList);
    this._updateUVOptions(options, stdMat, objDefs, false);
+   options.clearCoat = !!stdMat.clearCoat;
+   options.clearCoatGlossiness = !!stdMat.clearCoatGlossiness;
-   options.clearCoat = stdMat.clearCoat;
-   options.clearCoatGlossiness = stdMat.clearCoatGlossiness;
};
```

### PR DESCRIPTION
**Pull request by scarletsky**
Fixes #3189 

I confirm I have read the [contributing guidelines](https://github.com/playcanvas/engine/blob/master/.github/CONTRIBUTING.md) and signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/scene/materials/standard-material-options-builder.js [hunk 1]
```diff
@@ -39,8 +39,8 @@ StandardMaterialOptionsBuilder.prototype.updateRef = function (options, device,
     options.hasTangents = objDefs && stdMat.normalMap && ((objDefs & SHADERDEF_TANGENTS) !== 0);
     this._updateLightOptions(options, stdMat, objDefs, sortedLights, staticLightList);
     this._updateUVOptions(options, stdMat, objDefs, false);
-    options.clearCoat = stdMat.clearCoat;
-    options.clearCoatGlossiness = stdMat.clearCoatGlossiness;
+    options.clearCoat = !!stdMat.clearCoat;
+    options.clearCoatGlossiness = !!stdMat.clearCoatGlossiness;
```
Comment (mvaligursky) [inline src/scene/materials/standard-material-options-builder.js:43]: I wonder if those should be moved to _updateMaterialOptions - that's where the other material options are.

Comment (scarletsky) [reply src/scene/materials/standard-material-options-builder.js]: I just move the `clearCoat` and `clearCoatGlossiness` to `_updateMaterialOptions`.

**Review by mvaligursky — APPROVED** [review]

**Comment by mvaligursky** [comment]
assigned to @raytranuk to merge in when all approvals are done.

**Comment by willeastcott** [comment]
Thanks a lot for this! ❤️

**Pull request merged**